Review followup: PR #19309 — t2139: review-bot-gate require positive completion signal in bot_has_real_review

## Unaddressed review bot suggestions

PR #19309 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/19309

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `prompts/build.txt`
section 6a). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `prompts/build.txt` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/review-bot-gate-helper.sh:171`
- `.agents/scripts/tests/test-review-bot-gate-completion-signal.sh:115`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/review-bot-gate-helper.sh:171`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/19309#discussion_r3094695905)

```diff
+	# Mirrors _get_rate_limit_behavior — same repos.json schema, same precedence.
+	local repo_slug="$1"
+	local bot_login="$2"
+	local repos_json="${HOME}/.config/aidevops/repos.json"
+
+	if [[ -f "$repos_json" ]] && command -v jq &>/dev/null; then
+		local lag=""
+		lag=$(jq -r --arg slug "$repo_slug" --arg bot "$bot_login" \
+			'first(.initialized_repos[] | select(.slug == $slug)) | (.review_gate.tools[$bot].min_edit_lag_seconds // .review_gate.min_edit_lag_seconds // empty)' \
+			"$repos_json" 2>/dev/null) || lag=""
+		# Reject non-integer or negative values silently (fall through to env).
+		if [[ -n "$lag" && "$lag" != "null" && "$lag" =~ ^[0-9]+$ ]]; then
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The check `"$lag" != "null"` is redundant here. When using `jq -r` with a fallback like `// empty` or `// ""`, a `null` value in JSON results in an empty string in the shell variable. A non-empty check (`[[ -n "$lag" ]]`) is sufficient to handle missing or null values.
> 
> ```suggestion
> if [[ -n "$lag" && "$lag" =~ ^[0-9]+$ ]]; then
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. When parsing a JSON value in a shell script using `jq -r '.key // ""'`, the resulting variable will be an empty string if the key is missing or its value is `null`. Therefore, a check for the literal string "null" is redundant, and a non-empty check (e.g., `[[ -n "$var" ]]`) is sufficient.
> </details>

#### gemini-code-assist on `.agents/scripts/tests/test-review-bot-gate-completion-signal.sh:115`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/19309#discussion_r3094695933)

```diff
+load_helper_functions() {
+	# shellcheck disable=SC1090
+	# Trick: source the helper but make main() a no-op so it doesn't error.
+	# We strip the trailing 'main "$@"' invocation by sourcing in a subshell
+	# alternative — actually, the helper has `main "$@"` at the end which
+	# would execute. We override main BEFORE sourcing? No — sourcing runs
+	# top-level immediately. Solution: source via a temp file with the last
+	# `main "$@"` line removed.
+	local tmpfile
+	tmpfile=$(mktemp)
+	# Drop the final main invocation line (last non-empty line is `main "$@"`).
+	sed '$d' "$HELPER_SCRIPT" >"$tmpfile"
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> Using `sed '$d'` to remove the `main` call is fragile as it assumes the call is on the very last line and that there are no trailing newlines or comments. A more robust approach is to filter out the specific line using `grep`.
> 
> ```suggestion
> grep -v '^main "$@"' "$HELPER_SCRIPT" >"$tmpfile"
> ```

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/19309#pullrequestreview-4122519715)

> ## Code Review
> 
> This pull request implements a hybrid classification system for bot reviews to prevent premature merges caused by placeholder comments, specifically targeting CodeRabbit's two-phase posting pattern. It introduces a "settled" check that requires comments to be either edited after a specific duration or old enough to be considered final, alongside an expanded set of non-review patterns. Configuration for minimum edit lag has been added at the global, per-repo, and per-tool levels. Review feedback suggests simplifying a redundant null check when processing jq output and using a more robust method for stripping the main function call during test setup.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.8.59 automated scan.

--- a/.agents/scripts/review-bot-gate-helper.sh
+++ b/.agents/scripts/review-bot-gate-helper.sh
@@ -168,7 +168,7 @@ _get_min_edit_lag() {
 			'first(.initialized_repos[] | select(.slug == $slug)) | (.review_gate.tools[$bot].min_edit_lag_seconds // .review_gate.min_edit_lag_seconds // empty)' \
 			"$repos_json" 2>/dev/null) || lag=""
 		# Reject non-integer or negative values silently (fall through to env).
-		if [[ -n "$lag" && "$lag" != "null" && "$lag" =~ ^[0-9]+$ ]]; then
+		if [[ -n "$lag" && "$lag" =~ ^[0-9]+$ ]]; then
 			printf '%s' "$lag"
 			return 0
 		fi

--- a/.agents/scripts/tests/test-review-bot-gate-completion-signal.sh
+++ b/.agents/scripts/tests/test-review-bot-gate-completion-signal.sh
@@ -111,8 +111,8 @@ load_helper_functions() {
 	# `main "$@"` line removed.
 	local tmpfile
 	tmpfile=$(mktemp)
-	# Drop the final main invocation line (last non-empty line is `main "$@"`).
-	sed '$d' "$HELPER_SCRIPT" >"$tmpfile"
+	# Drop the final main invocation line using grep for robustness.
+	grep -v '^main "\$@"' "$HELPER_SCRIPT" >"$tmpfile"
 	# shellcheck disable=SC1090
 	source "$tmpfile"
 	rm -f "$tmpfile"

